lib/Sema/DerivedConformanceComparable.cpp

```diff
@@ -0,0 +1,411 @@
+//===--- DerivedConformanceComparable.cpp - Derived Comparable -===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2017 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+//  This file implements implicit derivation of the Comparable protocol.
+//  (Most of this code is similar to code in `DerivedConformanceEquatableHashable.cpp`)
+//
+//===----------------------------------------------------------------------===//
+
+#include "TypeChecker.h"
+#include "swift/AST/Decl.h"
+#include "swift/AST/Stmt.h"
+#include "swift/AST/Expr.h"
+#include "swift/AST/Module.h"
+#include "swift/AST/Pattern.h"
+#include "swift/AST/ParameterList.h"
+#include "swift/AST/ProtocolConformance.h"
+#include "swift/AST/Types.h"
+#include "llvm/ADT/APInt.h"
+#include "llvm/ADT/SmallString.h"
+#include "llvm/Support/raw_ostream.h"
+#include "DerivedConformances.h"
+
+using namespace swift;
+
+/// Returns a generated guard statement that checks whether the given lhs and
+/// rhs expressions are equal. If not equal, the else block for the guard
+/// returns lhs < rhs.
+/// \p C The AST context.
+/// \p lhsExpr The first expression to compare for equality.
+/// \p rhsExpr The second expression to compare for equality.
+static GuardStmt *returnComparisonIfNotEqualGuard(ASTContext &C,
```

Comment:
Same formatting nit as below. Please adhere to LLVM style consistently.

Reply:
This function generalizes one in the Equatable-Hashable code.  It can also be promoted to private API on `DerivedConformance` to offset the duplication.

Reply:
refactored into 3 separate functions to reduce code duplication
```
  static 
  GuardStmt *returnIfNotEqualGuard(ASTContext &C, Expr *lhsExpr, Expr *rhsExpr, 
    Expr *guardReturnValue);
  // return false 
  static 
  GuardStmt *returnFalseIfNotEqualGuard(ASTContext &C, Expr *lhsExpr, Expr *rhsExpr);
  // return lhs < rhs
  static 
  GuardStmt *returnComparisonIfNotEqualGuard(ASTContext &C, Expr *lhsExpr, Expr *rhsExpr);
```